spyrmsd/rmsd.py

```diff
@@ -373,3 +373,181 @@ def rmsdwrapper(
             )
 
     return RMSDlist
+
+
+import os
+from concurrent.futures import ProcessPoolExecutor
+from functools import partial
+from multiprocessing import Process, Queue
+
+
+def _rmsd_queue(
+    molref: molecule.Molecule,
+    mols: Union[molecule.Molecule, List[molecule.Molecule]],
+    queue: Queue,
+    symmetry: bool = True,
+    center: bool = False,
+    minimize: bool = False,
+    strip: bool = True,
+    cache: bool = True,
+) -> Any:
+    """
+    Compute RMSD between two molecules and put it in a queue.
```

Comment:
Is this computed lazily, or in the queue there is the actual result (computed)?

Reply:
No it is actually put in a queue, after which it is read immediately. If I could I wouldn't even have that `_rmsd_queue` function, but I couldn't find a reliable way to get the actual return output when using a Process, and using a queue seemed to be the most robust option.. It would have been a lot cleaner to just directly use the rmsdwrapper function in the timeout function process.

Reply:
Ok, so the way it works is that if the computation of `rmsdwrapper` takes a long time, when `process.join(timeout=timeout)` is called the `_rmsd_queue` has not completed yet and therefore the exit code is non-zero?

I'm not sure I fully understand the use of `Queue` here. I assume it's thread-safe since it comes from multi-processing, but there is only a single call to `rmsdwrapper` with all the molecules, right? Why a simple variable initialised to `None` would not work in this case? Maybe I'm missing something obvious here, I think I need a bit more time to wrap my head around this.

Reply:
> Ok, so the way it works is that if the computation of `rmsdwrapper` takes a long time, when `process.join(timeout=timeout)` is called the `_rmsd_queue` has not completed yet and therefore the exit code is non-zero?

Initially it was checking if the process was still running. However I changed it to check the exit code instead. An exit code of 0 is only returned if the process ran successfully, which means it finished and ran without errors. If the process is still running or raised an error, we terminate the process and return the `np.nan` list. 
Now that I think about it, i'm not sure what happens if an error is raised and we try to terminate the process then. Maybe I should add another check to ensure the process is still alive before terminating?
```python
if process.is_alive():
    process.terminate()
```

> I'm not sure I fully understand the use of `Queue` here. I assume it's thread-safe since it comes from multi-processing, but there is only a single call to `rmsdwrapper` with all the molecules, right? Why a simple variable initialised to `None` would not work in this case? Maybe I'm missing something obvious here, I think I need a bit more time to wrap my head around this.

Yeah, I agree your proposed solution of simply passing a variable would be ideal. The problem I ran into is the communication between processes. Since this spawns a separate process, variables that are returned here are not available to the main process. The usage of the queue is the best way I found to get around this. I don't think it's elegant and if we can do without it would be great (and make everything a lot cleaner). I will try to experiment a bit more tomorrow!

Reply:
>  Since this spawns a separate process, variables that are returned here are not available to the main process.

Thanks for the clarification, I did not think of that and now everything makes sense, thanks! I guess we need the shared state if we want to implement the timeout.

Reading the documentation [Sharing state between processes](https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes), I think in this case using `multiprocessing.Vale` would be a valid alternative? I think it would be also better to name the variable `result` (or similar) instead of `queue`.

Reply:
Resolving this conversation because everything is clear now, thanks for the explanations!